File: n_array/python/n_array.py

from __future__ import annotations

from typing import List

class NArray:
    def __init__(self, cols: int, rows: int):
        self._content = [ None for _ in range(cols * rows) ]
        self._cols = cols
        self._rows = rows

    def get(self, i: int, j: int) -> int:
        return self._content[i * self._cols + j]

    def set(self, i: int, j: int, v: int) -> None:
        self._content[i * self._cols + j] = v

    @property
    def cols(self) -> int:
        return self._cols

    @property # 裝飾子 decorator
    def rows(self) -> int:
        return self._rows

def dump(board) -> None:
    print(f"盤面 (board) 內容：")

    for j in range(board.rows):
        for i in range(board.cols):
            print(f"  ({i}, {j}) = {board.get(i, j)}")

if __name__ == "__main__":
    board = NArray(3, 3)

    print(f"盤面 (board) 設定：")

    for j in range(board.rows):
        for i in range(board.cols):
            print(f"  ({i}, {j}) = {i + j * board.cols}")
            board.set(i, j, i + j * board.cols)

    dump(board)

# n_array.py

File: n_array/c++/n_array.cpp

#include <iostream>

using namespace std;

class NArray {
    public:
        const int COLS;
        const int ROWS;

    private:
        int *content_;

    public:
        NArray(int, int);
        ~NArray();

    public:
        int get(int, int);
        void set(int, int, int);
};

NArray::NArray(int cols, int rows) : COLS(cols), ROWS(rows) {
    content_ = new int [cols * rows];
}

NArray::~NArray() {
    delete[] content_;
}

int NArray::get(int i, int j) {
    return content_[i + j * COLS];
}

void NArray::set(int i, int j, int value) {
    content_[i + j * COLS] = value;
}

void dump(NArray *board) {
    cout << "盤面 (board) 內容：" << endl;

    for (int j = 0; j < board->ROWS; j ++) {
        for (int i = 0; i < board->COLS; i ++) {
            cout << "  (" << i << ", " << j << ") = "
                 << board->get(i, j) << endl;
        }
    }
}

int main(int argc, char** argv) {
    NArray *board = new NArray(3, 3);

    cout << "盤面 (board) 設定：" << endl;

    for (int j = 0; j < board->ROWS; j ++) {
        for (int i = 0; i < board->COLS; i ++) {
            cout << "  (" << i << ", " << j << ") = "
                 << i + j * board->COLS <<endl;

            board->set(i, j, i + j * board->COLS);
        }
    }

    dump(board);
}

// n_array.cpp
